### PR TITLE
Harden composability contracts and real-method execution

### PR DESCRIPTION
**Pull request by Xmaster6y**
## Summary

- replace exact stage-type coalescing with an explicit shared-execution protocol
- let compatible probing stages share one model pass while preserving independent artifacts
- treat incompatible effects as planner split boundaries instead of rejecting valid workflows
- harden hook cleanup, model restoration, and composite-context bookkeeping across reuse and failures
- correct weight-adapter artifact mapping, adapter argument filtering, and the public star-import surface
- make the documented capability matrix executable through real-method conformance tests

## Why

The existing planner introduced capability-driven execution, but the library still had gaps between the composability claim and observable behavior: real built-in stages did not exercise same-run coexecution, effect conflicts rejected whole workflows, nested artifact namespaces could not be extended, and several lifecycle/API edge cases were uncovered by end-to-end composition tests.

This change makes the supported boundary explicit and evidence-backed for ActivationCaching, IntegratedGradients, LRP, Probing, and Adapters. Unsupported or stateful combinations remain conservatively split.

## Impact

Users can now build multi-stage pipelines with stable named artifacts, extend TensorDict namespaces safely, and coexecute compatible probing stages in one model pass. Cleanup and restoration remain reliable when hooks fail or contexts are reused. The documentation now distinguishes workflow composition from same-run optimization and points to the tests that substantiate each supported method.

## Validation

- `just tests`: 468 passed, 97.49% coverage
- `pre-commit run --all-files`
- `just docs`
- `uv lock --check`
- `git diff --check`
- `uv run python -c 'from tdhook import *'`

### CODE REVIEW

File: src/tdhook/artifacts.py [hunk 1]
```diff
@@ -162,12 +162,12 @@ def attribution_adapter(result_key: ArtifactKey = "attr") -> ArtifactAdapter:
     )
 
 
-def weight_adapter(cache_key: ArtifactKey = "cache") -> ArtifactAdapter:
-    """Adapter contract for the existing :class:`tdhook.weights.Adapters` method."""
+def weight_adapter(output_key: ArtifactKey = "output") -> ArtifactAdapter:
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply src/tdhook/artifacts.py]: Addressed in f241683. `weight_adapter(cache_key=...)` is accepted as a deprecated alias, retains the corrected `outputs/model` contract, and has regression coverage including the conflicting-keyword case.
